### ISSUE TITLE
field_reassign_with_default on struct with `#[non_exhaustive]` or private fields is false positive

### ISSUE THREAD

**Issue by bryanburgers** [rust-lang/rust-clippy#6559]
Lint name: `field_reassign_with_default`

Given a library that contains a struct with a `#[non_exhaustive]` attribute on it:

```rust
#[derive(Default)]
#[non_exhaustive]
pub struct TestStruct {
    pub one: usize,
    pub two: usize,
}
```

Code like this causes a clippy warning:

```rust
fn main() {
    let mut t = the_lib::TestStruct::default();
    t.one = 1;
    t.two = 2;
    println!("{} {}", t.one, t.two);
}
```

Clippy would prefer to see the following code, which is a compiler error.

```rust
fn main() {
    let t = the_lib::TestStruct {
        one: 1,
        two: 2,
        ..the_lib::TestStruct::default()
    };
    println!("{} {}", t.one, t.two);
}
```

```
cannot create non-exhaustive struct using struct expression
```

Note that due to the rules of `#[non_exhaustive]` application, this is only an error when the struct is defined in a different crate (which is why I'm not able to provide a playground link).

> The attribute #[non_exhaustive], when attached to a struct or the variant of an enum, will prevent code **outside of the crate** defining it from constructing said struct or variant.
>
> — [Rust 1.40.0 release notes](https://blog.rust-lang.org/2019/12/19/Rust-1.40.0.html)

### Meta

- `cargo clippy -V`: `clippy 0.0.212 (e1884a8e 2020-12-29)`
- `rustc -Vv`:
  ```
  rustc 1.49.0 (e1884a8e3 2020-12-29)
  binary: rustc
  commit-hash: e1884a8e3c3e813aada8254edfa120e85bf5ffca
  commit-date: 2020-12-29
  host: x86_64-apple-darwin
  release: 1.49.0
  ```

**Comment by bryanburgers** [comment]
Hmm... actually this affects structs within the same crate that have private fields, as well.

```rust
mod the_mod {
    #[derive(Default)]
    pub struct OtherTestStruct {
        pub one: usize,
        pub two: usize,
        three: usize,
    }
}

fn main() {
    // Clippy warning
    let mut t = the_mod::OtherTestStruct::default();
    t.one = 1;
    t.two = 2;
    println!("{} {}", t.one, t.two);

    // Compiler error
    /*
    let t = the_mod::OtherTestStruct {
        one: 1,
        two: 2,
        ..the_mod::OtherTestStruct::default()
    };
    println!("{} {}", t.one, t.two);
    */
}
```

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=993abe21260563a8c5a12c4383a54d11
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=93692333032b262ba980aaaa311726d6

**Comment by bryanburgers** [comment]
Relevant rust issue: rust-lang/rust#70564
Related clippy issue: rust-lang/rust-clippy#6312

**Comment by Manishearth** [comment]
Fixed in https://github.com/rust-lang/rust-clippy/pull/6375
